packages/techdocs-common/src/stages/publish/googleStorage.ts

```diff
@@ -0,0 +1,222 @@
+    // and the storage client will work.
+    storageClient
+      .bucket(bucketName)
+      .getMetadata()
+      .then(() => {
+        logger.info(
+          `Successfully connected to the GCS bucket ${bucketName} in the GCP project ${projectId}.`,
+        );
+      })
+      .catch(reason => {
+        logger.error(
+          `Could not retrieve metadata about the GCS bucket ${bucketName} in the GCP project ${projectId}. ` +
+            'Make sure the GCP project and the bucket exists and the access key located at the path ' +
+            "techdocs.publisher.google.pathToKey defined in app config has the role 'Storage Object Creator'. " +
+            'Refer to https://backstage.io/docs/features/techdocs/using-cloud-storage',
+        );
+        throw new Error(`from GCS client library: ${reason.message}`);
+      });
+
+    return new GoogleGCSPublish(storageClient, bucketName, logger);
+  }
+
+  constructor(
+    private readonly storageClient: Storage,
+    private readonly bucketName: string,
+    private readonly logger: Logger,
+  ) {
+    this.storageClient = storageClient;
+    this.bucketName = bucketName;
+    this.logger = logger;
+  }
+
+  /**
+   * Upload all the files from the generated `directory` to the GCS bucket.
+   * Directory structure used in the bucket is - entityNamespace/entityKind/entityName/index.html
+   */
+  publish({ entity, directory }: PublishRequest): Promise<{}> {
+    return new Promise((resolve, reject) => {
+      // Path of all files to upload, relative to the root of the source directory
+      // e.g. ['index.html', 'sub-page/index.html', 'assets/images/favicon.png']
+      const allFilesToUpload: Array<string> = [];
+
+      // Iterate on all the files in the directory and its sub-directories
+      walk(directory)
+        .on('data', (item: walk.Item) => {
+          // GCS manages creation of parent directories if they do not exist.
+          // So collecting path of only the files is good enough.
+          if (item.stats.isFile()) {
+            // Remove the absolute path prefix of the source directory
+            const relativeFilePath = item.path.replace(`${directory}/`, '');
+            allFilesToUpload.push(relativeFilePath);
+          }
+        })
+        .on('error', (err: Error, item: walk.Item) => {
+          const errorMessage = `Unable to read file at ${item.path}. Error ${err.message}`;
+          this.logger.error(errorMessage);
+          reject(errorMessage);
+        })
+        .on('end', () => {
+          // 'end' event happens when all the files have been read.
+          const entityRootDir = `${entity.metadata.namespace}/${entity.kind}/${entity.metadata.name}`;
+          allFilesToUpload.forEach(filePath => {
+            const source = path.join(directory, filePath); // Local file absolute path
+            const destination = `${entityRootDir}/${filePath}`; // GCS Bucket file relative path
+            this.storageClient
```

Comment:
You don't await this, seems like an omission. It isn't blocking, is it?

Reply:
Indeed.  It's kind of nice right now, because I think this kicks off all uploads simultaneously (as-written), which is great for performance, but **a)** could become a problem due to API throttling on sufficiently large docs sites, and **b)** makes the `resolve()` below a little disingenuous (since it will resolve long before the files are actually on GCS).  This method may behave in unexpected ways overall too, since the promise may see several rejections.

Would recommend, for now, using [Promise.allSettled()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled) and then doing one big resolve/reject at the end, depending on the outcome.

A future performance/reliability enhancement could involve a library like Bluebird to throttle simultaneous uploads.

Reply:
We use `p-limit` in the catalog - ultra basic, but gets things done, for throttling.

Reply:
Thanks for the discussion on this! As Eric suggested, doing a `Promise.allSettled()` on all the promises returned by individual `.upload()`s will resolve "(b)" i.e. will make the the final `resolve/reject` meaningful.

About the API throttling, GCP seems to allow 1000 file uploads per second, but can be increased for buckets. https://cloud.google.com/storage/quotas#objects

Will add a note here to upload in chunks of 1000.

Reply:
It's not realistic to upload 1k wide concurrently from a single node. This will definitely hurt more than help. Limiting to something more like 10 or whatever would be more reasonable.